### ISSUE TITLE
`_calculate_cosine_importance` includes padding tokens in cosine similarity, distorting layer importance scores

### ISSUE THREAD

**Issue by peremartra** [peremartra/optipfair#38]
**Title:** `_calculate_cosine_importance` includes padding tokens in cosine similarity, distorting layer importance scores

## Summary

The `_calculate_cosine_importance` function used in `analyze_layer_importance` flattens hidden states to `[batch_size, seq_len * hidden_dim]` and computes cosine similarity per batch sample. Padding tokens are included in this computation. Their hidden states are numerically meaningless, but they dominate the flattened vector — especially with short sequences padded to a long `max_length` — producing artificially inflated or deflated importance scores that misidentify which layers are safe to prune.

## Root cause

In `_calculate_cosine_importance`, the tensors are flattened before computing similarity:

```python
input_flat = input_tensor.view(input_tensor.size(0), -1)
output_flat = output_tensor.view(output_tensor.size(0), -1)
similarity = F.cosine_similarity(input_valid, output_valid, dim=1)
```

The function does not accept an `attention_mask` parameter and has no way to exclude padding positions. The caller that runs the dataloader loop also does not extract or forward the mask.

## Why this matters

With `padding='max_length'` and short sequences, the padding fraction per batch can be very high. For example, with `max_length=512` and SMS-style messages (~15–20 tokens), over 96% of each flattened vector corresponds to padding hidden states. The similarity score is effectively measuring how the model processes padding, not real content. This can cause:

- Layers that are genuinely important for real tokens to appear passive (high similarity over padding)
- Layers that are genuinely passive to appear active
- Data-driven pruning decisions based on `analyze_layer_importance` to be systematically biased

The effect is more severe with short-text datasets and large `max_length` values, but it is present to some degree in any setup with variable-length sequences.

## Proposed fix

Add `attention_mask=None` as a trailing parameter to `_calculate_cosine_importance` to preserve full backward compatibility. When `None`, the function behaves exactly as today. When provided, it computes cosine similarity token-by-token (`dim=-1`), zeroes out padding positions, and averages only over valid tokens.

The internal caller that iterates the dataloader should also be updated to extract `attention_mask` from each batch and forward it.

## Regression test invariant

With proper masking, the importance score for any layer must be stable regardless of how much padding surrounds identical content tokens. A batch of short real sequences padded to `max_length=512` should produce the same scores as the same sequences with no padding.
